explainer.md

```diff
@@ -71,56 +64,32 @@ for all new messages including group chats not directly addressed to the user.)
 
 There is a single global badge associated with each Web
 application (as defined in [Web app
-manifest](https://www.w3.org/TR/appmanifest/)). At any time, the badge is set to
-either:
-
-* Nothing (the badge is "cleared"), or
-* A "flag" indicating the presence of a badge with no contents, or
-* A non-empty string containing a single *[grapheme
-  cluster](http://unicode.org/reports/tr29/#Grapheme_Cluster_Boundaries)*
-  (roughly: a single Unicode character), or
-* A positive integer.
-
-The model does not allow a badge that is the empty string, a negative integer,
-or the integer value 0.
+manifest](https://www.w3.org/TR/appmanifest/)). At any time, the badge is set with:
 
-The grapheme cluster allows a single character to be represented, but also
-for combining characters to be combined into what the user thinks of as a single
-character. Examples of grapheme clusters include:
-
-* "a" (a single character, Latin "a")
-* "நி" (a base character + combining character, Tamil "ni")
-* 👩🏾‍🔬️ (emoji combination with skin tone modifier, Woman Scientist with
-  Medium-Dark Skin Tone)
+* Nothing (we display a notification "flag"), or
+* A positive integer.
 
-Clients are encouraged to use integers instead of digit characters, because some
-operating systems may *only* support integer badges, so using the digit
-character would not display correctly. Also integers can be localized better for
-the user.
+The model does not allow a badge that is a negative integer or the integer value 0.
```

Comment:
> The model does not allow a badge that is a negative integer or the integer value 0.

Hmm interesting. I could see this making iterating and calculating total badge counts a little difficult when some of them are `flag` strings vs positive integers. Was there any specific reason for not allowing 0 for computation purposes?

Reply:
Do you mean for the case

```js
window.Badge.set(getUnreadCount())
```

when there are no unread emails? That's a good point. 

Do you think `Badge.set(0)` should be analogous to calling `Badge.clear()`? It raises the question, do we really need a `Badge.clear()` method at all?

It also means `set` will have quite a spread of different behaviors. 
`Badge.set()` --> Sets the badge to be a flag (perhaps a white/red dot)
`Badge.set(0)` --> Clears the badge.
`Badge.set(number)` --> Sets the badge to be the number (falling back to the white/red dot on platforms that don't support it).

But it probably would be easier than having to write something like this
```js
const unreadCount = getUnreadCount();
if (unreadCount)
    window.Badge.set(unreadCount);
else
    window.Badge.clear()
```

Reply:
See this more precise (but not yet updated) description in the spec draft:
https://wicg.github.io/badging/#set-method

My plan was for `set(0)` (and `set("")` before we removed strings) to be equivalent to `clear()`. As I acknowledged in the issue box, it's a bit weird that `set(0)` would clear the badge, while `set()` would set the flag. But the reason I did this is precisely so you don't have to write that if statement.

I don't understand the original concern that this "[makes] iterating and calculating total badge counts a little difficult", unless you just mean the above if statement. We don't want to be showing "0" on all apps that have unread messages. Setting the count to 0 should just clear the badge.

> It raises the question, do we really need a `Badge.clear()` method at all?

It's redundant, but I think it's still useful, since the only other way to clear a badge would be `set(0)` which is a little unintuitive. If you had an app that just wants to show a Boolean flag, it's unintuitive that you would call `set()` to set the flag and `set(0)` to clear it. The idea is that you use the API in one of two different ways:

1. Always call `set(number)`, e.g. with your unread count. If the number is 0, it gets cleared.
2. Call `set()` and `clear()` without arguments, to turn it on and off.

The other thing I would consider is removing `clear` and removing the empty argument to `set`, and introducing a Boolean argument to `set`. So now the rules are:

- `set(number)` sets a positive integer. 0 clears.
- `set(bool)` sets a flag if true, clears if false.

Reply:
+1 to reducing overloads and having a separate method for the flag-version. The design is clever, but especially in a language without a compiler to enforce valid usage, avoiding overloads is probably the better approach.

Reply:
(I'm happy to continue discussing this here, but I won't block the PR on it since this is off-topic from the matter at hand.)

- I don't necessarily see the API as "incomplete" without a getter. This API is essentially a fire and forget / best effort output to the OS. You can think of it a bit like a print function: you're sending output; if you want to know what output you sent, you need to remember it yourself. The fact that we allow the UA to "forget" the badge contents at any time means that if we _did_ add a getter, you could use it to determine whether the UA has forgotten about your badge, but you wouldn't be able to rely on this for incrementing. You should keep your own count if you want that. (Basically, a getter adds more complexity.)
- I don't think of this as an overloaded function. I see it as a data model that can be set to multiple types. Having separate functions for updating the same data model feels even more confusing to me. For example, if you call `set(3)`, then `hide()`, then later `show()`, does that re-show the "3" or is it just displaying the flag status? I don't think of what I've got here as particularly "clever". It's the most honest representation of the data model we've got underneath. (The Boolean option I mentioned above is possibly an even more honest representation.)

Reply:
Thanks for your response @mgiuca. I realize there is quite a bit to unpack in my response above. So I've moved the discussions over to #18 and #19. We can continue discussions there.

@domenic, your point about overloading the `set()` is valid and I think we should continue discussion on that. Do you think we should create another issue for that?